### PR TITLE
C# autobuilder: Build all solutions at the highest depth

### PR DESCRIPTION
**Pull request by lcartey**
One of the most common reasons for the C# autobuilder to fail on customer code is because it cannot determine a single unique .sln or .proj file to build. Instead, if multiple sln or proj files appear at the same shortest depth from the root, the autobuilder currently logs an error, and doesn't try to build any of them.

This proposed change instead will trigger building of all the solutions or project files at the shortest depth. This is a backwards compatible change, in the sense that anything that succeeded with the autobuilder before will proceed in the same way, but we will now also build repositories without a single unique solution/project file.

Target of  `rc/1.22`, as this is an issue that occurs relatively frequently at customers.

### CODE REVIEW

**Comment by lcartey** [comment]
Added a test case.

File: csharp/autobuilder/Semmle.Autobuild.Tests/BuildScripts.cs [hunk 1]
```diff
@@ -745,6 +745,54 @@ public void TestWindowCSharpMsBuild()
             TestAutobuilderScript(autobuilder, 0, 6);
         }
 
+	[Fact]
```
Comment (calumgrant) [inline csharp/autobuilder/Semmle.Autobuild.Tests/BuildScripts.cs]: Indentation

File: csharp/autobuilder/Semmle.Autobuild.Tests/BuildScripts.cs [hunk 2]
```diff
@@ -745,6 +745,54 @@ public void TestWindowCSharpMsBuild()
...
+            Actions.RunProcess[@"cmd.exe /C C:\odasa\tools\csharp\nuget\nuget.exe restore test2.csproj"] = 0;
+            Actions.RunProcess["cmd.exe /C CALL ^\"C:\\Program Files ^(x86^)\\Microsoft Visual Studio 12.0\\VC\\vcvarsall.bat^\" && C:\\odasa\\tools\\odasa index --auto msbuild test2.csproj /p:UseSharedCompilation=false /t:Windows /p:Platform=\"x86\" /p:Configuration=\"Debug\" /p:MvcBuildViews=true /P:Fu=Bar"] = 0;
+            Actions.RunProcess[@"cmd.exe /C C:\odasa\tools\java\bin\java -jar C:\odasa\tools\extractor-asp.jar ."] = 0;
+            Actions.RunProcess[@"cmd.exe /C C:\odasa\tools\odasa index --xml --extensions config csproj props xml"] = 0;
+            Actions.FileExists["csharp.log"] = true;
+            Actions.FileExists[@"test1.csproj"] = true;
+            Actions.FileExists[@"test2.csproj"] = true;
+	    Actions.FileExists[@"C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe"] = false;
```
Comment (calumgrant) [inline csharp/autobuilder/Semmle.Autobuild.Tests/BuildScripts.cs]: Indentation

File: csharp/autobuilder/Semmle.Autobuild.Tests/BuildScripts.cs [hunk 3]
```diff
@@ -745,6 +745,54 @@ public void TestWindowCSharpMsBuild()
...
+    </ItemGroup>
+  </Project>");
+            Actions.LoadXml["test2.csproj"] = csproj2;
+
+            var autobuilder = CreateAutoBuilder("csharp", true, msBuildArguments: "/P:Fu=Bar", msBuildTarget: "Windows", msBuildPlatform: "x86", msBuildConfiguration: "Debug",
+                vsToolsVersion: "12");
+
+            TestAutobuilderScript(autobuilder, 0, 6);
+        }
+
```
Comment (calumgrant) [inline csharp/autobuilder/Semmle.Autobuild.Tests/BuildScripts.cs]: Extra newline

**Comment by lcartey** [comment]
Done.

**Comment by jf205** [comment]
Thanks for the mention @hvitved 

> We will need to update the C# auto builder documentation accordingly.

Do you mean this page https://lgtm.com/help/lgtm/csharp-extraction?

**Comment by hvitved** [comment]
> Do you mean this page https://lgtm.com/help/lgtm/csharp-extraction?

Yes, specifically
```
Invoke dotnet build on the solution or project file closest to the top level directory.
Invoke msbuild on the solution or project file closest to the top level directory.
```
should now be (I guess)
```
Invoke dotnet build on the solution(s) or project file(s) closest to the top level directory.
Invoke msbuild on the solution(s) or project file(s) closest to the top level directory.
```

**Comment by felicitymay** [comment]
This also needs a change note so that it's included in the 1.22.1 release notes.

**Pull request merged**